### PR TITLE
Benchmark grdsample, grdfilter and sph2grd with fixed cores

### LINKED ISSUE #2942: Investigate large (>10%) deviations in benchmark of test_grdlandmask_no_outgrid
### Description of the problem

The [`test_grdlandmask_no_outgrid`](https://github.com/GenericMappingTools/pygmt/blob/5639f24d7a511493330830585b9c2eff919a1cf0/pygmt/tests/test_grdlandmask.py#L47-L58) function's benchmark was added in #2911, but it seems like the performance has been deviating wildly (>10%) in many PRs since (e.g. https://github.com/GenericMappingTools/pygmt/pull/2937#issuecomment-1873549552, https://github.com/GenericMappingTools/pygmt/pull/2938#issuecomment-1873554070), even though there has been no evident change in the code of `grdlandmask` or any of the underlying clib functions.

See tracked performance at https://codspeed.io/GenericMappingTools/pygmt/benchmarks/pygmt/tests/test_grdlandmask.py::test_grdlandmask_no_outgrid

Sample flame graph from https://codspeed.io/GenericMappingTools/pygmt/branches/windows-multiprocessing:

image

Opening this issue to discuss why the variance might be so high, and if there are ways to mitigate this.

### Minimal Complete Verifiable Example

```python
make test PYTEST_EXTRA="-r P --pyargs pygmt -k test_grdlandmask_no_outgrid --codspeed"
```

### Full error message

_No response_

### System information

```bash
PyGMT information:
    version: v0.10.1.dev187+g322f8889
  System information:
    python: 3.12.1 | packaged by conda-forge | (main, Dec 23 2023, 08:03:24) [GCC 12.3.0]
    executable: /usr/share/miniconda/bin/python
    machine: Linux-6.2.0-1018-azure-x86_64-with-glibc2.35
  Dependency information:
    numpy: 1.26.2
    pandas: 2.1.4
    xarray: 2023.12.0
    netCDF4: 1.6.5
    packaging: 23.2
    contextily: None
    geopandas: 0.14.1
    ipython: None
    rioxarray: None
    ghostscript: 10.02.1
  GMT library information:
    binary version: 6.4.0
    cores: 4
    grid layout: rows
    image layout: 
    library path: /usr/share/miniconda/lib/libgmt.so
    padding: 2
    plugin dir: /usr/share/miniconda/lib/gmt/plugins
    share dir: /usr/share/miniconda/share/gmt
    version: 6.4.0
```

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

Benchmarking OpenMP multi-threaded algorithms with a fixed number of CPU cores, to ensure consistency of benchmarks.

Using a fixed core count of 2, which is the maximum number of CPU cores for GitHub Actions' Linux runners - https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

TODO:
- [x] [test_grdsample_dataarray_out](https://codspeed.io/GenericMappingTools/pygmt/benchmarks/pygmt/tests/test_grdsample.py::test_grdsample_dataarray_out)
- [x] [test_sph2grd_no_outgrid](https://codspeed.io/GenericMappingTools/pygmt/benchmarks/pygmt/tests/test_sph2grd.py::test_sph2grd_no_outgrid)
- [ ] [test_grdimage_grid_and_shading_with_xarray[png]](https://codspeed.io/GenericMappingTools/pygmt/benchmarks/pygmt/tests/test_grdimage.py::test_grdimage_grid_and_shading_with_xarray%5Bpng%5D) (errors, see https://github.com/GenericMappingTools/pygmt/pull/2945#discussion_r1441213340)
- [x] [test_grdfilter_dataarray_in_dataarray_out](https://codspeed.io/GenericMappingTools/pygmt/benchmarks/pygmt/tests/test_grdfilter.py::test_grdfilter_dataarray_in_dataarray_out)

Fixes #2942

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/tests/test_grdimage.py [hunk 1]
```diff
@@ -139,10 +139,10 @@ def test_grdimage_grid_and_shading_with_xarray(grid, xrgrid):
     """
     fig_ref, fig_test = Figure(), Figure()
     fig_ref.grdimage(
-        grid="@earth_relief_01d_g", region="GL", cmap="geo", shading=xrgrid
+        grid="@earth_relief_01d_g", region="GL", cmap="geo", shading=xrgrid, cores=2
     )
```
Comment (weiji14) [inline pygmt/tests/test_grdimage.py:143]: This line is raising an error at https://github.com/GenericMappingTools/pygmt/actions/runs/7404780202/job/20146784435?pr=2945#step:8:778:

```
 E           pygmt.exceptions.GMTCLibError: Module 'grdimage' failed with status code 72:
Error:      grdimage [ERROR]: Option -x given more than once
Error:      grdimage [ERROR]: Option -x parsing failure. Correct syntax:
Error:      grdimage [ERROR]: Offending option -x2
```

Equivalent GMT CLI command also fails:

```
gmt grdimage @earth_relief_01d_g -RGL -Cgeo -x2 -png map
```

```
grdimage [ERROR]: Option -x given more than once
grdimage [ERROR]: Option -x parsing failure. Correct syntax:

   [-x[[-]<n>]] .
     Limit the number of cores used in multi-threaded algorithms [Default uses all 16 cores]. If <n> is negative then we select (16 - <n>) cores (or at least 1).
grdimage [ERROR]: Offending option -x2
```

Might be an upstream GMT bug?

File: .github/workflows/benchmarks.yml [hunk 2]
```diff
@@ -92,6 +92,7 @@ jobs:
         with:
           run: |
             python -c "import pygmt; pygmt.show_versions()"
+            gmt set GMT_MAX_CORES=2
```
Comment (seisman) [inline .github/workflows/benchmarks.yml:95]: remove this line?

File: pygmt/tests/test_grdimage.py (hunk 1 above)
Comment (seisman) [reply pygmt/tests/test_grdimage.py]: very likely

Comment (weiji14) [reply pygmt/tests/test_grdimage.py]: Ok, reported upstream at https://github.com/GenericMappingTools/gmt/issues/8256. I've reverted the `grdimage` commit at 08f36d2b3eaa5ba6181412afbc0af0b2dc56bdcd for now.

File: .github/workflows/benchmarks.yml (hunk 2 above)
Comment (weiji14) [reply .github/workflows/benchmarks.yml]: Ok, done at 31dce8e1740de6c1353324e8fbaffaae66e93e4e

**Review by seisman — APPROVED** [review]

**Pull request merged**